### ISSUE TITLE
Anchor links to same page shows as broken

### ISSUE THREAD

**Issue by amolswnz** [silverstripe/silverstripe-elemental#894]
## Issue:
Anchor links on internal or same page shows as broken.

### My dev setup:
I am able to replicate this issue on basic Silverstripe site
My composer.json
```
    "require": {
        "php": "^7.1 || ^8",
        "silverstripe/recipe-plugin": "^1.2",
        "silverstripe/recipe-cms": "4.7.3@stable",
        "silverstripe-themes/simple": "~3.2.0",
        "silverstripe/login-forms": "4.3.0@stable",
        "dnadesign/silverstripe-elemental": "^4.5"
    },
```
### To Replicate:
- In existing or new content block, create an anchor tag
- Click Insert Link -> Anchor on page
- Chose above anchor tag
- Save/Publish Page
- New Tag is generated as `<p><a class="ss-broken" href="[sitetree_link,id=1]#anchor">Link to some anchor</a></p>`

### Related
https://github.com/silverstripe/silverstripe-elemental/issues/796

### ACs
- [ ] Broken anchor on existing pages are not reported as broken links

### PRs
- https://github.com/silverstripe/silverstripe-elemental/pull/911

### Note
- The [current logic for identifying broken anchor](https://github.com/silverstripe/silverstripe-cms/blob/4/code/Model/SiteTreeLinkTracking_Parser.php#L72) is very weak and will report a lot of false positive aside from Elemental blocks. Given that, we think it's better to remove it and only report anchor link broken if the parent page is missing all together

**Comment by emteknetnz** [comment]
Thanks for reporting Amol.

Could I get you to quickly find the exact versions of cms and elemental that you're running?  You can do that with
`composer show | grep '/cms\|/silverstripe-elemental'`

Cheers

**Comment by amolswnz** [comment]
```
dnadesign/silverstripe-elemental          4.5.0   Elemental pagetype and collection of Elements
silverstripe/cms                          4.7.1   The SilverStripe Content Management System
```

**Comment by KhushbuFuletra** [comment]
Hi 
We are having similar issues.  Anchor links on the same page show as broken. Any updates on this?

Just adding my findings below if that helps in debugging the issue further.
 
Anchors work fine when you select a different page on the "Link to an anchor on a page" popup modal form, but not when you select the same page. (Though doesn't throw any error in the console.)
However, AnchorSelectField::getAnchorsInPage()  method does _not_ seem to get called on the initial form load; I'm guessing that the initial set of information in the AnchorSelectField is just loaded from the current WYSIWYG field only and **not** the whole page.

Adding below list of element module versions and their dependencies.

```
dnadesign/silverstripe-elemental                             4.0.5                                  Elemental pagetype and collection of Elements
dnadesign/silverstripe-elemental-list                       1.2.0                                  Adds a new element for nested elements
dnadesign/silverstripe-elemental-userforms            3.0.0                                  Adds a new element for usersforms
silverstripe/cms                                                        4.7.1                                  The SilverStripe Content Management System
symbiote/silverstripe-elemental-dashboard             1.0.x-dev 7e14217             Dashboard module with three column elemental area layout
thewebmen/silverstripe-elemental-grid                    2.x-dev dc60207               Elemental grid module
```
Cheers
Khushbu

**Comment by emteknetnz** [comment]
Core issue is that SiteTreeLinkTracking_Parser (which is used by the SiteTreeLinkTracking DataExension, which DOES run on elemental blocks) assumes that page content is $page->content, while elemental content is $page .. ElementalAreas() .. Elements() ... HTMLText fields (which are also used in templates)

```php
} elseif (!empty($matches['anchor'])) {
    // Ensure anchor isn't broken on target page
    $anchor = preg_quote($matches['anchor'], '/');
    $broken = !preg_match("/(name|id)=\"{$anchor}\"/", $page->Content); // <<< here's your issue
} 
```

https://github.com/silverstripe/silverstripe-cms/blob/4/code/Model/SiteTreeLinkTracking_Parser.php#L72

Seems like need to add quite a bit of API to do this properly, something along these lines:

```php
// SiteTree.php
public function getRendereredContent(): string
{
    $renderedContent = [];
    $this->updateRenderedContent($renderedContent);
    $this->extend('augmentRenderedContent');
}

// protected to allow subclasses to override this i.e. NOT include $page->Content on BlockPage
protected function updateRenderedContent(array &$renderedContent)
{
    $renderedContent[] = $this->Content;
}

// BlockPage.php
protected function updateRenderedContent(array &$renderedContent)
{
   // noop
}

// ElementalPageExtension.php
public function augmentRenderedContent(array &$renderedContent)
{
    foreach ([ElementalAreas] as $area) {
        foreach ($area->Elements() as $element) {
           foreach ($element->[HTMLText fields] as $field) {
              // this is assuming that ALL HTMLText fields are rendered in templates, should probably have a private static array to specific what will is rendered
              $renderedContent[] = $element->$field;
           }
        }
    } 
}
```

This would allow use to update SiteTreeLinkTracking_Parser to use $page->getRenderedContent() to look for broken links

**Comment by brynwhyman** [comment]
It's perhaps also worth noting there's the dream of a single 'content' API: https://github.com/silverstripe/silverstripe-cms/issues/2454

**Comment by emteknetnz** [comment]
Yup, that single content API is pretty much what's required here

**Comment by GuySartorelli** [comment]
The root problem here is that anchors inside an elemental block are not properly supported. This will be resolved with #186, so I'm closing this issue in favour of that one.
